src/main/java/org/kizombadev/markdownparser/SyntaxAnalyzer.java

```diff
@@ -0,0 +1,161 @@
+package org.kizombadev.markdownparser;
+
+import com.google.common.collect.ImmutableList;
+import org.kizombadev.markdownparser.entities.*;
+
+import java.util.Stack;
+
+public class SyntaxAnalyzer {
+
+    private ImmutableList<Token> tokens;
+    private int tokenIndex = 0;
+    private Stack<State> stack = new Stack<>();
+
+    public static SyntaxAnalyzer create() {
+        return new SyntaxAnalyzer();
+    }
+
+    public Syntax parse(ImmutableList<Token> tokens) {
+        this.tokens = tokens;
+
+        RootSyntax root = new RootSyntax();
+
+        while (currentToken() != null) {
+            if (currentToken().equals(Token.NumberSign)) {
+                handleBigHeadlineLine(root);
+            } else if (currentToken().equals(Token.DoubleNumberSign)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') NullPointerException might be thrown as 'currentToken' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS2259)